File: Problem11/zigzagite.cpp
#include<bits/stdc++.h>
struct TreeNode
{
     int val;
     TreeNode *left,*right;
};

class Solution {
public:
    std::vector<std::vector<int>>ans;
    void levelorderite(TreeNode *root)
    {
        if(root==NULL)
             return ;
        int val=1,sz;
        std:: deque<TreeNode*>dq;
        TreeNode *cur;
        std::vector<int>temp;
        dq.push_front(root);
        while(!dq.empty())
        {
            sz=dq.size();
            temp.clear();
            for(int i=0;i<sz;i++)
            {
                if(val==1)
                {
                    cur=dq.front();
                    dq.pop_front();
                    if(cur->left!=NULL)
                           dq.push_back(cur->left);
                    if(cur->right!=NULL)
                           dq.push_back(cur->right);
                }
                else
                {
                    cur=dq.back();
                    dq.pop_back();
                    if(cur->right!=NULL)
                          dq.push_front(cur->right);
                    if(cur->left!=NULL)
                          dq.push_front(cur->left);
                }
                temp.push_back(cur->val);
            }
            ans.push_back(temp);
            val^=1;
            
        }
    }
    std::vector<std::vector<int>> zigzagLevelOrder(TreeNode* root) {
        ans.clear();
        levelorderite(root);
        return ans;
    }
};

File: Problem11/zigzagite.py
from ast import List
from logging.config import valid_ident
from typing import Optional
import collections
class TreeNode:
    def __init__(self,val=0,left=None,right=None) -> None:
        self.val=val
        self.left=left
        self.right=right
class Solution:
    def zigzagite(root)->List[List[int]]:
        ans=[]
        if(root==None):
            return ans
        dq=collections.deque()
        dq.append(root)
        val=1
        cur=None
        while len(dq)>0:
            sz=len(dq)
            temp=[]
            for i in range(0,sz):
                if val==1:
                    cur=dq[0]
                    dq.popleft()
                    if(cur.left!=None):
                         dq.append(cur.left)
                    if(cur.right!=None):
                        dq.append(cur.right)
                else:
                    cur=dq[-1]
                    dq.pop()
                    if(cur.right!=None):
                        dq.appendleft(cur.right)
                    if(cur.left!=None):
                        dq.appendleft(cur.left)
                temp.append(cur.val)
            ans.append(temp)
            val=1-val
        return ans
                        
                   
      
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        return Solution.zigzagite(root)
       